src/Avalonia.Ide.CompletionEngine/XmlParser.cs

```diff
@@ -29,17 +31,52 @@ public enum ParserState
         private int? _attributeNameEnd;
         private int _attributeValueStart;
         private Stack<int> _containingTagStart;
+        private bool _isClosingTag;
 
-        public string TagName => State >= ParserState.StartElement
-            ? _data.Substring(_elementNameStart, (_elementNameEnd ?? (_data.Length - 1)) - _elementNameStart + 1)
-            : null;
+        public string TagName
+        {
+            get
+            {
+                if (State < ParserState.StartElement)
+                {
+                    return null;
+                }
+                var span = _data.Span;
+                if(_elementNameStart >= span.Length)
+                {
+                    return "";
+                }
+
+                if (_elementNameEnd == null)
+                {
+                    // Check where is element end
+                    _elementNameEnd = _elementNameStart;
```

Comment:
Why are we mutating the parser state in property getter?

Reply:
I've modified parser a bit to allow handling spans in longer strings.
Now there is possibility, to ask for TagName of tag, which is not fully parsed. This code performs look-ahead to return whole TagName, no matter where parser is stopped.

What do you think is best solution here, do not mutate state and calculate end position each time someone is asking for TagName?

Reply:
That might require adjustments for the rest of the completion engine since it expects to get the unfinished name in the `TagName` property.

Reply:
But `XmlParser` only parses span passed to `Parse` method. In completion engine it only gets text up to cursor position, so it will not be affected.
I do not want to change way of `CompletionEngine` works atm, I think it's too much effort without clear value. I plan to pass full text to CompletionEngine for [this](https://github.com/kekekeks/Avalonia.Ide/pull/23), but Xml will still be parsed up to cursor position and additional checks will be performed on raw text.

Reply:
I think we need two separate properties, e. g.  `FullTagName` and `TagNameUpToCursor` or something like that to better describe the meaning of the property. `FullTagName` shouldn't mutate the state.

I think that we might need the same change for other properties as well.

Reply:
Maybe `GetFullTagName` method, which would not mutate state?

Reply:
+

Reply:
Done